lib/jsdom/living/helpers/http-request.js

```diff
@@ -0,0 +1,272 @@
+"use strict";
+const http = require("http");
+const https = require("https");
+const { Writable } = require("stream");
+const zlib = require("zlib");
+
+const ver = process.version.replace("v", "").split(".");
+const majorNodeVersion = Number.parseInt(ver[0]);
+const minorNodeVersion = Number.parseInt(ver[1]);
+
+function abortRequest(clientRequest) {
+  if (majorNodeVersion > 13 || (majorNodeVersion === 13 && minorNodeVersion > 13)) {
+    clientRequest.destroy();
+  } else {
+    clientRequest.abort();
+  }
+  clientRequest.removeAllListeners();
+  clientRequest.on("error", () => {});
+}
+
+class Request extends Writable {
+  constructor(uriHref, options) {
+    super();
+    this._currentUrl = uriHref;
+    this._options = options;
+    this.headers = this._options.headers;
+    this._ended = false;
+    this._ending = false;
+    this._redirectCount = 0;
+    this._requestBodyLength = 0;
+    this._requestBodyBuffers = [];
+    this.response = Object.create(null);
+    this._performRequest();
+  }
+
+  abort() {
+    abortRequest(this._currentRequest);
+    this.emit("abort");
+    this.removeAllListeners();
+  }
+
+  getCookie() {
+    let cookies;
+    if (this._redirectCount === 0) {
+      this.originalCookieHeader = this.getHeader("Cookie");
+    }
+    const jar = this._options.cookieJar;
+    if (jar && typeof jar.getCookieStringSync === "function") {
+      cookies = jar.getCookieStringSync(this._currentUrl);
+    } else {
+      this._areCookiesDisabled = true;
+    }
+    if (cookies && cookies.length) {
+      if (this.originalCookieHeader) {
+        this.setHeader("Cookie", this.originalCookieHeader + "; " + cookies);
+      } else {
+        this.setHeader("Cookie", cookies);
+      }
+    }
+  }
+
+  setCookies(res) {
+    const self = this;
+    const jar = self._options.cookieJar;
+    const cookies = res.headers["set-cookie"];
+    if (!this._areCookiesDisabled && jar && typeof jar.setCookieSync === "function" && Array.isArray(cookies)) {
+      try {
+        cookies.forEach(cookie => jar.setCookieSync(cookie, self._currentUrl, { ignoreError: true }));
+      } catch (e) {
+        this.emit("error", e);
+      }
+    }
+  }
+
+  write(data, encoding) {
+    if (this._ending) {
+      const writeAfterEndError = new Error("Writing data after ClientRequest.end() is called is not allowed");
+      writeAfterEndError.code = "ERR_STREAM_WRITE_AFTER_END";
+      throw writeAfterEndError;
+    }
+    if (!(typeof data === "string" || (typeof data === "object" && ("length" in data)))) {
+      throw new TypeError("data should be a string, Buffer or Uint8Array");
```

Comment:
Since we are presumably the only people writing here, could we make this always use a single type (e.g. Uint8Array)?

Also, this kind of explicit type-checking is not desired, since again we control the code so we're just checking ourselves.